core/trino-main/src/main/java/io/trino/execution/DropSchemaTask.java

```diff
@@ -72,10 +74,33 @@ public String explain(DropSchema statement, List<Expression> parameters)
             return immediateVoidFuture();
         }
 
+        if (!isSchemaEmpty(session, schema, metadata)) {
+            throw semanticException(SCHEMA_NOT_EMPTY, statement, "Cannot drop non-empty schema '%s'", schema.getSchemaName());
+        }
+
         accessControl.checkCanDropSchema(session.toSecurityContext(), schema);
 
         metadata.dropSchema(session, schema);
 
         return immediateVoidFuture();
     }
+
+    private static boolean isSchemaEmpty(Session session, CatalogSchemaName schema, Metadata metadata)
+    {
+        QualifiedTablePrefix tablePrefix = new QualifiedTablePrefix(schema.getCatalogName(), schema.getSchemaName());
+
+        if (!metadata.listTables(session, tablePrefix).isEmpty()) {
```

Comment:
Worth adding a comment that's this is best effort check, and doesn't provide any guarantees in the face of concurrent operations.